File: IN3130/Oblig2/src/python/flow_test.py
import numpy as np
import os

from Flow import Flow


def read_graph_from_file(file_path: str) -> np.ndarray:
    with open(file_path, 'r') as f:
        N = int(f.readline().strip())
        graph = np.zeros((N, N), dtype=int)
        for i in range(N):
            graph[i] = list(map(int, f.readline().strip().split()))
    return graph


class TestFlow:
    def test_flow_case1(self):
        graph = read_graph_from_file('../test/resources/assignment/testcase1.txt')
        print(graph)
        flow_solver = Flow(graph)
        max_flow = flow_solver.solve()
        assert max_flow == 24
        assert flow_solver.get_cut() == [0, 2]

    def test_flow_case2(self):
        graph = read_graph_from_file('../test/resources/assignment/testcase2.txt')
        flow_solver = Flow(graph)
        max_flow = flow_solver.solve()
        assert max_flow == 40
        assert flow_solver.get_cut() == [0, 1, 4, 5, 6, 7, 8, 9, 10, 11, 12]

    def test_flow_case3(self):
        graph = read_graph_from_file('../test/resources/assignment/testcase3.txt')
        flow_solver = Flow(graph)
        max_flow = flow_solver.solve()
        assert max_flow == 98921
        assert flow_solver.get_cut() == [0, 7, 10]

    def test_flow_case4(self):
        graph = read_graph_from_file('../test/resources/assignment/testcase4.txt')
        flow_solver = Flow(graph)
        max_flow = flow_solver.solve()
        assert max_flow == 84
        assert flow_solver.get_cut() == [0, 4, 5, 6, 7, 8, 9, 11, 13, 14, 15, 22, 25, 26, 27, 29, 31, 32, 34, 36, 37]

File: IN3130/Oblig2/src/test/java/assignment/FlowTest.java
package assignment;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertIterableEquals;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;

public class FlowTest {

    @Test
    public void testFlowCase1() {
        int[][] graph = readGraphFromFile("/assignment/testcase1.txt");
        Flow flowSolver = new Flow(graph);
        int maxFlow = flowSolver.solve();
        assertEquals(24, maxFlow);

        List<Integer> cut = flowSolver.cut();
        assertIterableEquals(Arrays.asList(0, 2), cut);
    }

    @Test
    public void testFlowCase2() {
        int[][] graph = readGraphFromFile("/assignment/testcase2.txt");
        Flow flowSolver = new Flow(graph);
        int maxFlow = flowSolver.solve();
        assertEquals(40, maxFlow);

        List<Integer> cut = flowSolver.cut();
        assertIterableEquals(Arrays.asList(0, 1, 4, 5, 6, 7, 8, 9, 10, 11, 12), cut);
    }

    @Test
    public void testFlowCase3() {
        int[][] graph = readGraphFromFile("/assignment/testcase3.txt");
        Flow flowSolver = new Flow(graph);
        int maxFlow = flowSolver.solve();
        assertEquals(98921, maxFlow);

        List<Integer> cut = flowSolver.cut();
        assertIterableEquals(Arrays.asList(0, 7, 10), cut);
    }

    @Test
    public void testFlowCase4() {
        int[][] graph = readGraphFromFile("/assignment/testcase4.txt");
        Flow flowSolver = new Flow(graph);
        int maxFlow = flowSolver.solve();
        assertEquals(84, maxFlow);

        List<Integer> cut = flowSolver.cut();
        assertIterableEquals(Arrays.asList(0, 4, 5, 6, 7, 8, 9, 11, 13, 14, 15, 22, 25, 26, 27, 29, 31, 32, 34, 36, 37), cut);
    }

    private int[][] readGraphFromFile(String filePath) {
        try (BufferedReader reader = new BufferedReader(
                new InputStreamReader(Objects.requireNonNull(FlowTest.class.getResourceAsStream(filePath))))) {
            int N = Integer.parseInt(reader.readLine());
            int[][] graph = new int[N][N];

            for (int i = 0; i < N; i++) {
                String[] line = reader.readLine().split(" ");
                for (int j = 0; j < N; j++) {
                    graph[i][j] = Integer.parseInt(line[j]);
                }
            }
            return graph;
        } catch (Exception e) {
            throw new RuntimeException("Error reading graph from file", e);
        }
    }
}
